std/experimental/checkedint.d

```diff
@@ -0,0 +1,2509 @@
+}
+
+unittest
+{
+    alias opCmpProper = ProperCompare.hookOpCmp;
+    assert(opCmpProper(42, 42) == 0);
+    assert(opCmpProper(42u, 42) == 0);
+    assert(opCmpProper(42, 42u) == 0);
+    assert(opCmpProper(-1, uint(-1)) < 0);
+    assert(opCmpProper(uint(-1), -1) > 0);
+    assert(opCmpProper(-1.0, -1) == 0);
+}
+
+unittest
+{
+    auto x1 = Checked!(uint, ProperCompare)(42u);
+    assert(x1.get < -1);
+    assert(x1 > -1);
+}
+
+// WithNaN
+/**
+
+Hook that reserves a special value as a "Not a Number" representative. For
+signed integrals, the reserved value is `T.min`. For signed integrals, the
+reserved value is `T.max`.
+
+The default value of a $(D Checked!(X, WithNaN)) is its NaN value, so care must
+be taken that all variables are explicitly initialized. Any arithmetic and logic
+operation involving at least on NaN becomes NaN itself. All of $(D a == b), $(D
+a < b), $(D a > b), $(D a <= b), $(D a >= b) yield `false` if at least one of
+`a` and `b` is NaN.
+
+*/
+struct WithNaN
+{
+static:
+    /**
+    The default value used for values not explicitly initialized. It is the NaN
+    value, i.e. `T.min` for signed integrals and `T.max` for unsigned integrals.
+    */
+    enum T defaultValue(T) = T.min == 0 ? T.max : T.min;
+    /**
+    The maximum value representable is $(D T.max) for signed integrals, $(D
+    T.max - 1) for unsigned integrals. The minimum value representable is $(D
+    T.min + 1) for signed integrals, $(D 0) for unsigned integrals.
+    */
+    enum T max(T) = cast(T) (T.min == 0 ? T.max - 1 : T.max);
+    /// ditto
+    enum T min(T) = cast(T) (T.min == 0 ? T(0) : T.min + 1);
+
+    /**
+    If `rhs` is `WithNaN.defaultValue!Rhs`, returns
+    `WithNaN.defaultValue!Lhs`. Otherwise, returns $(D cast(Lhs) rhs).
+
+    Params:
+    rhs = the value being cast (`Rhs` is the first argument to `Checked`)
+    Lhs = the target type of the cast
+
+    Returns: The result of the cast operation.
+    */
+    Lhs hookOpCast(Lhs, Rhs)(Rhs rhs)
+    {
+        static if (is(Lhs == bool))
+        {
+            return rhs != defaultValue!Rhs && rhs != 0;
+        }
+        else static if (valueConvertible!(Rhs, Lhs))
+        {
+            return rhs != defaultValue!Rhs ? Lhs(rhs) : defaultValue!Lhs;
+        }
+        else
+        {
+            if (isUnsigned!Rhs || !isUnsigned!Lhs ||
+                    Rhs.sizeof > Lhs.sizeof || rhs >= 0)
+            {
+                auto result = cast(Lhs) rhs;
+                // If signedness is different, we need additional checks
+                if (result == rhs &&
+                        (!isUnsigned!Rhs || isUnsigned!Lhs || result >= 0))
+                    return result;
+            }
+            return defaultValue!Lhs;
+        }
+    }
+
+    /**
+    Unconditionally returns `WithNaN.defaultValue!Lhs` for all `opAssign`
+    failures.
+
+    Params:
+    Lhs = The target of the assignment (`Lhs` is the first argument to
```

Comment:
There is no Lhs parameter. Perhaps you want to document that just the type of the second parameter is used.